### PR TITLE
Fixed issue #18832: Participants can't re-enter a submitted survey if setting is using "inherit" at participant settings

### PR DESCRIPTION
**Pull request by gabrieljenik**
- Adding oOptions to the conditions
- Placing some conditions as model methods (this is to start giving entity to some business rules)

### CODE REVIEW

**Review by Shnoulle — APPROVED** [review]
:+1:

**Comment by Shnoulle** [comment]
**Not related to this issue** : Looking at code i think we must check Valid DFrom and Valid Until too

If someone want to check and report.

File: application/models/TokenDynamic.php [hunk 1]
```diff
@@ -1105,4 +1105,30 @@ public function getSurveyId()
     {
         return self::$sid;
     }
+
+    /**
+     * Returns true if the token can be used
+     * @return bool
+     */
+    public function canBeUsed()
```
Comment (Shnoulle) [inline application/models/TokenDynamic.php:1113]: :+1:

File: application/models/TokenDynamic.php [hunk 2]
```diff
@@ -1105,4 +1105,30 @@ public function getSurveyId()
...
+    {
+        return !empty($this->token)
+            && (
+                $this->completed == "N"
+                || empty($this->completed)
+                || $this->survey->oOptions->alloweditaftercompletion == "Y"
+            );
+    }
+
+    public function canBeEmailed()
```
Comment (Shnoulle) [inline application/models/TokenDynamic.php:1123]: :+1:

File: application/models/TokenDynamic.php [hunk 3]
```diff
@@ -1105,4 +1105,30 @@ public function getSurveyId()
...
+    {
+        return !empty($this->token)
+            && $this->emailstatus == "OK"
+            && $this->email
+            && $this->completed == "N"
```
Comment (Shnoulle) [inline application/models/TokenDynamic.php:1128]: `( $this->completed == "N" || empty($this->completed)))`

Same than can be used

File: application/models/TokenDynamic.php [hunk 4]
```diff
@@ -1105,4 +1105,30 @@ public function getSurveyId()
...
+    /**
+     * Returns true if the token can be used
+     * @return bool
+     */
+    public function canBeUsed()
+    {
+        return !empty($this->token)
+            && (
+                $this->completed == "N"
+                || empty($this->completed)
```
Comment (Shnoulle) [inline application/models/TokenDynamic.php:1118]: No `$this->usesleft > 0 `check ? (same than email)

**Review by Shnoulle — APPROVED** [review]
OK, but maybe some test improvment.

**Comment by gabrieljenik** [comment]
> OK, but maybe some test improvment.

Yes, will add them later on another PR

File: application/models/TokenDynamic.php (hunk 4 above)
Comment (gabrieljenik) [reply application/models/TokenDynamic.php:1118]: That was ported from the gridButtons part.
I think that could be part of `canBeUsed` but have some doubts about scenarios, ex: editing a prior response.
Anyway, don't want to update anything now, jusr reog a bit the code to start cetralizing business rules.

File: application/models/TokenDynamic.php (hunk 3 above)
Comment (gabrieljenik) [reply application/models/TokenDynamic.php:1128]: Yes, there seems to be some relationship. So far, don't want to make further changes until we have clear scenarios and better tests in place

**Comment by Shnoulle** [comment]
> > OK, but maybe some test improvment.
> 
> Yes, will add them later on another PR

My question are something between : «Need to report a DEV issue on mantis ?» and «Maybe we can fix it quickly now»

No strong opinion. approved the update ;)

**Review by Shnoulle — APPROVED** [review]
better !

**Pull request merged**